Refactor unit tests and fix ability to use EventBus without runtime transformers

LexManos opened the pull request:
Reinstate our old non-ModLoader ASM Event Handler creator.
Add default isCancelable && hasResult implementations that do not rely on transforming Event subclasses.
This reinstated EventBus's ability to be used as a normal library without Runtime class transformations.
Changed BusBuilder to an interface, with a private implementation for clener API.

AterAnimAvis commented on eventbus-test/src/test/java/net/minecraftforge/eventbus/test/TestModLauncherBase.java:
```diff
@@ -0,0 +1,87 @@
+    private static final String METHOD_NAME = "test.modlauncher.method";
+
+    void validate(Class<?> clazz) {
+        // We expect transformers to not run, so make sure LISTENER_LIST does not exist
```
Comment doesn't match following check, seems to be inverted

LexManos replied:
Ya, copy pasted comment from NoLoader whops.

AterAnimAvis commented on eventbus-testjars/src/main/java/net/minecraftforge/eventbus/benchmarks/compiled/BenchmarkArmsLength.java:
```diff
@@ -16,8 +14,11 @@
-            if (!new CancelableEvent().isCancelable())
-                throw new RuntimeException("Transformer did not apply!");
+            try {
+                Event.class.getDeclaredField("LISTENER_LIST");
```
This change currently breaks JMH, currently JMH applies transformations over the TestJar, EventBus itself isn't being transformed so this check fails.

LexManos replied:
Ya JMH's task was today. I just wanted to push a sanity check of the test project before I went to bed.

AterAnimAvis commented on src/main/java/net/minecraftforge/eventbus/EventBus.java:
```diff
@@ -147,6 +155,12 @@ private void registerListener(final Object target, final Method method, final Me
+        {
+            throw new IllegalArgumentException("Failed to create ASMEventHandler for " + target.getClass().getName() + "." + method.getName() + Type.getMethodDescriptor(method) + " it is not public and our transformer is disabled");
+        }
+        System.out.println(target.getClass().getName() + "." + method.getName() + Type.getMethodDescriptor(method) + " " + method.getModifiers());
```
Rogue `System.out.println`

LexManos replied:
What? No! You're the one who left debug lines in >.>

AterAnimAvis reviewed (changes requested):
The MockLauncherHandlerService in JMH can also be removed as well (mirroring the removal in the tests)

LexManos commented:
```
Benchmark                                              Mode  Cnt    Score    Error  Units
EventBusBenchmark.testClassLoaderCombined              avgt   15  187.639 ▒  4.703  ns/op
EventBusBenchmark.testClassLoaderDynamic               avgt   15   47.457 ▒  2.715  ns/op
EventBusBenchmark.testClassLoaderLambda                avgt   15   48.543 ▒  1.034  ns/op
EventBusBenchmark.testClassLoaderStatic                avgt   15   44.938 ▒  0.417  ns/op
EventBusBenchmark.testModLauncherCombined              avgt   15  199.128 ▒  7.225  ns/op
EventBusBenchmark.testModLauncherDynamic               avgt   15   49.537 ▒  2.950  ns/op
EventBusBenchmark.testModLauncherLambda                avgt   15   48.335 ▒  0.506  ns/op
EventBusBenchmark.testModLauncherStatic                avgt   15   44.696 ▒  1.219  ns/op
EventBusBenchmarkNoLoader.testNoLoaderCombined         avgt   15  443.092 ▒ 10.731  ns/op
EventBusBenchmarkNoLoader.testNoLoaderDynamic          avgt   15  220.423 ▒  6.489  ns/op
EventBusBenchmarkNoLoader.testNoLoaderLambda           avgt   15  226.937 ▒  3.677  ns/op
EventBusBenchmarkNoLoader.testNoLoaderStatic           avgt   15  216.448 ▒  5.148  ns/op
```
Fixed benchmarks. Using a transformer is SO much faster then not, as expected.

LexManos commented:
This should be ready to merge.
However I want someone else {preferably cpw} to read through he changes in `src/main` as that's actual changes modder facing.

Some notes:
EventSubclassTransformer: I added code to transform Event itself, if it ever comes up. To use faster variants of getListenerList, hasResult, and isCanceable. And because we typically can't transform Event itself. I also made anything that's a direct subclass of Event implement faster hasResult/isCancelable to bypass the map based lookup of that stuff. So cancelable/result work now in non-transformed environments. Its just slower obviously.

IEventListenerFactory: Added to allow us to use the old ASMClassCloader style of defining classes, meant to not be exposed publicly, Modders select what they want through BusBuilder.useModLauncher(). Using ModLauncher is now opt-in cuz I couldn't think of a non-hacky way to detect if ML is running.

LockHelper: I needed to use the same thread locking mechanics for the non-transformed implementation of isCancelaable and hasRecord, so I refactored this to a helper. Again not ment as public API.

Names: I got tired of my keyboard typoing magic strings in method names so I added the Method helper :D

Now for the important bit:
I changed BusBuilder to an interface instead of a final class and moved the implementation to the non-api package. This removed the getters, which were  used by the EventBus constructor. This *should* be source and binary compatible IF they didn't use the getters. But I haven't verified.

Event: Removed the deprecated getParentListenerList and Setup

Once cpw gives a review  of this and calls it good. I think we should do a pass to convert the license header to the compact format Forge now uses. As well as standardize it.

cpw commented:
```
Benchmark                                              Mode  Cnt    Score   Error  Units
EventBusBenchmark.testClassLoaderCombined              avgt   15  129.463 ± 3.729  ns/op
EventBusBenchmark.testClassLoaderCombined:·stack       avgt           NaN            ---
EventBusBenchmark.testClassLoaderDynamic               avgt   15   32.740 ± 0.261  ns/op
EventBusBenchmark.testClassLoaderDynamic:·stack        avgt           NaN            ---
EventBusBenchmark.testClassLoaderLambda                avgt   15   33.156 ± 0.375  ns/op
EventBusBenchmark.testClassLoaderLambda:·stack         avgt           NaN            ---
EventBusBenchmark.testClassLoaderStatic                avgt   15   30.865 ± 0.195  ns/op
EventBusBenchmark.testClassLoaderStatic:·stack         avgt           NaN            ---
EventBusBenchmark.testModLauncherCombined              avgt   15  128.407 ± 1.932  ns/op
EventBusBenchmark.testModLauncherCombined:·stack       avgt           NaN            ---
EventBusBenchmark.testModLauncherDynamic               avgt   15   32.491 ± 0.368  ns/op
EventBusBenchmark.testModLauncherDynamic:·stack        avgt           NaN            ---
EventBusBenchmark.testModLauncherLambda                avgt   15   33.237 ± 0.483  ns/op
EventBusBenchmark.testModLauncherLambda:·stack         avgt           NaN            ---
EventBusBenchmark.testModLauncherStatic                avgt   15   30.926 ± 0.292  ns/op
EventBusBenchmark.testModLauncherStatic:·stack         avgt           NaN            ---
EventBusBenchmarkNoLoader.testNoLoaderCombined         avgt   15  244.138 ± 6.832  ns/op
EventBusBenchmarkNoLoader.testNoLoaderCombined:·stack  avgt           NaN            ---
EventBusBenchmarkNoLoader.testNoLoaderDynamic          avgt   15  110.356 ± 0.804  ns/op
EventBusBenchmarkNoLoader.testNoLoaderDynamic:·stack   avgt           NaN            ---
EventBusBenchmarkNoLoader.testNoLoaderLambda           avgt   15  121.564 ± 1.323  ns/op
EventBusBenchmarkNoLoader.testNoLoaderLambda:·stack    avgt           NaN            ---
EventBusBenchmarkNoLoader.testNoLoaderStatic           avgt   15  110.712 ± 1.328  ns/op
EventBusBenchmarkNoLoader.testNoLoaderStatic:·stack    avgt           NaN            ---
```

cpw reviewed (approved):
If you think this replaces all the old tests fair enough. I don't have the time to validate that personally. The JMH runs fine for me and results are comparable between old and new.

LexManos commented:
Ya, If I understood all the old tests properly it replaces them 1:1, I added a couple extra because I wanted to verify my own things.
